### PR TITLE
Fix Internal Compiler Error on VS 2015.2

### PR DESCRIPTION
**Pull request by ThePhD**
This fixes the internal compiler error found in #7.

### CODE REVIEW

**Comment by DouglasHeriot** [comment]
Thanks! I ran into this too.
I had a quick try at fixing it, but couldn't work out what it was that actually made the compiler crash. Was there something in its error output that helped you work out a workaround?

**Comment by ThePhD** [comment]
Jawdropping amounts of experience getting variadics to behave nice with VC++ since the November 2012 CTP. At some point, you (unfortunately) begin to understand the compiler itself, because you eat so many ICEs a day.

**Comment by MyOwnClone** [comment]
Thanks! This solved my problem with ICEs in VS 2015.2!

File: include/misc/exception.hpp [hunk 1]
```diff
@@ -39,7 +39,7 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #else
 #   define EXCEPTION_HPP_NORETURN_MACRO [[noreturn]]
 #endif
-
+#if 0
```
Comment (Habbie) [inline include/misc/exception.hpp]: If we need this code nowhere, I suggest removing it. Otherwise, perhaps we need a more specific if than `0`?

**Comment by Habbie** [comment]
Hello, can you review my comment and rebase onto master so the PR gets tested? Thank you!

**Comment by stbuehler** [comment]
The changes to `exception.hpp` should probably be dropped; that part was fixed in #10.

File: include/LuaContext.hpp [hunk 2]
```diff
@@ -1615,11 +1615,10 @@ class LuaContext {
             throw WrongTypeException(lua_typename(state, index), typeid(TFirstType));
 
         Binder<TCallback, const TFirstType&> binder{ callback, *firstElem };
-        return readIntoFunction(state, retValueTag, binder, index + 1, othersTags...);
+        return readIntoFunction(state, retValueTag, binder, index + 1, tag<TTypes...>());
     }
-    template<typename TRetValue, typename TCallback, typename TFirstType, typename... TTypes>
-    static auto readIntoFunction(lua_State* state, tag<TRetValue> retValueTag, TCallback&& callback, int index, tag<TFirstType>, tag<TTypes>... othersTags)
-        -> typename std::enable_if<!IsOptional<TFirstType>::value, TRetValue>::type
```
Comment (stbuehler) [inline include/LuaContext.hpp]: Was there a reason why you moved the `enable_if`? `, TRetValue` should be unnecessary in this case, the `void` default is fine there.

Comment (ThePhD) [inline include/LuaContext.hpp]: I was not being too terribly thorough in my rewrite: the goal was to get it compiling and working so I could benchmark it against sol2.

File: include/misc/exception.hpp (hunk 1 above)
Comment (ThePhD) [inline include/misc/exception.hpp]: Probably something more specific than `0`. I don't know on which platforms this code was necessary, and I didn't end up needing it, and it was interfering with compilation.

**Comment by ThePhD** [comment]
This has been rebased by #26 and parts of it taken up by #10 as well.

**Pull request closed without merging**